### PR TITLE
Added unit tests for Parameters

### PR DESCRIPTION
**Pull request by mikejbrown**
It's good practice to start adding as much testing as possible. Here's a start on a particularly easy to test class: `Parameters`.

Uses Unity Editor tests based on NUnit. For those unfamiliar (assuming there are some newbies around) see the [Unity docs](https://docs.unity3d.com/Manual/testing-editortestsrunner.html) and [NUnit docs](http://www.nunit.org/index.php?p=quickStart&r=2.6.4). Run the tests in the Unity editor by going Window > Editor Tests Runner then clicking Run All.

Had to make some minor fixes to `Parameters` to get all the tests running. Some test(s) will have to be changed if #746 is merged.

From now on changes to the `Parameters` API can be spec'ed out in the tests in terms of the _expected_/_desired_ behavior, and _then_ the implementation coded to make the tests pass (test driven development).

### CODE REVIEW

File: Assets/Scripts/Models/Parameter.cs [hunk 1]
```diff
@@ -138,13 +138,13 @@ public void WriteXmlParamGroup(XmlWriter writer)
 
         writer.WriteStartElement("Param");
         writer.WriteAttributeString("name", name);
-        if (!value.Equals(""))
+        if (value != null && !value.Equals(""))
```
Comment (svmnotn) [inline Assets/Scripts/Models/Parameter.cs]: here you could just use `string.IsNullOrEmpty(value) == false`

Comment (mikejbrown) [inline Assets/Scripts/Models/Parameter.cs]: I was not aware of that function. Made the change. Thank you! :+1:

**Comment by koosemose** [comment]
From a read through, the tests look good... though seeing that I apparently wrapped what should be a variable in quotes... wow I was tired when I got to WriteXml...

**Comment by mikejbrown** [comment]
@koosemose We've all been there. That's why it's so important to write tests!!

File: Assets/Scripts/Models/Parameter.cs (hunk 1 above)
Comment (svmnotn) [inline Assets/Scripts/Models/Parameter.cs]: No problem, it is a very hidden very useful function.

File: Assets/Scripts/Models/Parameter.cs [hunk 2]
```diff
@@ -138,13 +138,13 @@ public void WriteXmlParamGroup(XmlWriter writer)
...
         writer.WriteAttributeString("name", name);
-        if (!value.Equals(""))
+        if (string.IsNullOrEmpty(value) == false)
         {
             writer.WriteAttributeString("value", value);
         }
         foreach (string k in contents.Keys)
         {
-            this["k"].WriteXml(writer);
+            this[k].WriteXml(writer);
```
Comment (Grenkin1988) [inline Assets/Scripts/Models/Parameter.cs]: k -> key

**Comment by vogonistic** [comment]
Merged! Thanks for starting the work with adding tests. I've figured out how to run them from the cli as well, but one of the tests fails now that we've fixed the exception.

From inside the ProjectPorcupine directory (with Unity closed): `/Applications/Unity/Unity.app/Contents/MacOS/Unity -batchmode -runEditorTests -quit -projectPath $PWD`

**Pull request merged**